# Secant method

The secant method fall between bisection and Newton methods as a way to solve 1d nonlinear equations $f(x)=0$. The idea is best explained with a picture.



In [1]:
using Plots
gr()

x = linspace(0.5,3.5)
f(x) = x.^2 - 4

a = 1
b = 3
fa = f(a)
fb = f(b)

c = a - fa*(b - a)/(fb - fa)

δ = 1 # a small displacement for ploting labels

plot(x, f(x), color="blue",  linestyle=:solid, label="", linewidth=2)
plot!(x, 0*x, color="black", linestyle=:solid, label="")
plot!([a, b], [fa, fb], color="blue", linestyle=:dash, label="")

scatter!([a, b],    [fa, fb],  marker=:circ, color="blue",  label="")
scatter!([a, b, c], [0, 0, 0], marker=:circ, color="black", label="")

annotate!(a, δ, text("a"))
annotate!(b, δ, text("b"))
annotate!(c, δ, text("c"))
annotate!(a, f(a)-δ, text("f(a)"))
annotate!(b, f(b)+1.3δ, text("f(b)"))
annotate!(3.6, 9, text("f(x)"))
plot!(xlabel="x", ylabel="y", ylims=(-5, 10), xlims=(0,4))


<?xml version="1.0" encoding="utf-8"?>
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 0 
 
 
 1 
 
 
 2 
 
 
 3 
 
 
 4 
 
 
 -5 
 
 
 0 
 
 
 5 
 
 
 10 
 
 
 x 
 
 
 y 
 
<polyline clip-path="url(#clip02)" style="stroke:#0000ff; stroke-width:2; stroke-opacity:1; fill:none" points="
 111.921,331.044 120.114,329.536 128.308,327.853 136.501,325.997 144.695,323.966 152.888,321.761 161.082,319.383 169.275,316.83 177.469,314.103 185.662,311.202 
 193.856,308.127 202.049,304.878 210.242,301.455 218.436,297.858 226.629,294.087 234.823,290.141 243.016,286.022 251.21,281.729 259.403,277.261 267.597,272.62 
 275.79,267.804 283.984,262.815 292.177,257.651 300.371,252.313 308.564,246.802 316.758,241.116 324.951,235.256 333.145,229.222 341.338,223.014 349.532,216.632 
 357.725,210.076 365.919,203.346 374.112,196.442 382.306,189.364 390.499,182.111 398.693,174.685 406.886,167.085 415.08,159.31 423.273,151.362 431.467,143.239 
 439.66,134.943 447.854,126.472 456.047,117.827 464.241,109.008 472.434,100.016 480.628,90.8488 488.821,81.5078 497.015,71.9928 505.208,62.3038 513.401,52.4407 
 
 "/>
<polyline clip-path="url(#clip02)" style="stroke:#000000; stroke-width:1; stroke-opacity:1; fill:none" points="
 111.921,243.981 120.114,243.981 128.308,243.981 136.501,243.981 144.695,243.981 152.888,243.981 161.082,243.981 169.275,243.981 177.469,243.981 185.662,243.981 
 193.856,243.981 202.049,243.981 210.242,243.981 218.436,243.981 226.629,243.981 234.823,243.981 243.016,243.981 251.21,243.981 259.403,243.981 267.597,243.981 
 275.79,243.981 283.984,243.981 292.177,243.981 300.371,243.981 308.564,243.981 316.758,243.981 324.951,243.981 333.145,243.981 341.338,243.981 349.532,243.981 
 357.725,243.981 365.919,243.981 374.112,243.981 382.306,243.981 390.499,243.981 398.693,243.981 406.886,243.981 415.08,243.981 423.273,243.981 431.467,243.981 
 439.66,243.981 447.854,243.981 456.047,243.981 464.241,243.981 472.434,243.981 480.628,243.981 488.821,243.981 497.015,243.981 505.208,243.981 513.401,243.981 
 
 "/>
 
 
 
 
 
 
 
 
 
 
 
 
 a 
 
 
 b 
 
 
 c 
 
 
 f(a) 
 
 
 f(b) 
 
 
 f(x)

We picture the secant method beginning $a,b$ bracketing a root, like the bisection method. Only instead of taking the midpoint as the next iterate, we draw the secant to the curve (the line that connects the two points $(a, f(a))$ and $(b, f(b))$), and we choose the point $c$ where the secant crosses the $y$ axis. 

Let's derive a formula for the value of $c$. Think of the secant line $\hat{f}(x)$ as a linear approximation to $f(x)$ that starts at point $(a, f(a))$ and moves upward from there with slope $(f(b) - f(a))/(b-a)$. The formula for such a line is

\begin{equation*}
\hat{f}(x) = f(a) + (x-a) \, \frac{f(b) - f(a)}{b-a}
\end{equation*}

At the intersection point with $y=0$ we have $\hat{f}(c) = 0$. Plug that in to the above

\begin{equation*}
0 = f(a) + (c-a) \, \frac{f(b) - f(a)}{b-a}
\end{equation*}

and solve for $c$ to get

\begin{equation*}
c = a - f(a) \, \frac{b-a}{f(b) - f(a)}
\end{equation*}

Unlike the bisection algorithm, $c$ is not guaranteed to lie between $a$ and $b$, so we need some other way of choosing which of $a$ and $b$ to replace for the iteration. Standard practice is to replace whichever of $|f(a)|$ and $|f(b)|$ is larger. 

## Pseudocode for secant algorithm

```
secantsearch(f, a, b)
    start with a,b preferably bracketing a root of f(x)
    while approximate solution is still not good enough
       compute c = a - f(a) (b-a)/(f(b) - f(a))    
       if |f(c)| is not smaller than the lesser of |f(a)| and |f(b)|
          then algorithm isn't converging, so exit
       end    
       if |f(a)| > |f(b)|
          let a = c
       else
          let b = c
       end      
    end      
    return c
end
```

## Julia implementation of secant algorithm

In [2]:
function secantsearch(f, a::AbstractFloat, b::AbstractFloat, diagnostics=false) 
# return solution x of f(x) == 0, with a,b serving as initial guesses
    
    c = (a+b)/2         # c will take on bigger float type of a,b
    fa = f(a)           # cache values of f(a), f(b), f(c)      
    fb = f(b)
    fc = f(c)
    ϵ = eps(typeof(c))  # machine epsilon
    n = 1               # iteration count
    N = 1024            # max iterations
    
    # stop when f(c) is small, |b-a| is small relative to a and b
    # or when we reach maximium number of iterations
    while abs(fc) > 10ϵ && abs(b-a)/(abs(a)+abs(b)) > 10ϵ && n < N
    
        # compute new values for c, fc by secant approximation
        c = a - fa*(b-a)/(fb-fa)
        fc = f(c)
        
        # print diagnostics if they're turned on
        diagnostics && println("n = $n, c = $c, f(c) = $(f(c))")
        
        # error and exit if f(c) is bigger than both f(a) and f(b)
        if abs(fc) > abs(fa) && abs(fc) > abs(fb)
            println("secantsearch error: f(c) is increasing")
            return c
        end
        
        # replace whichever of a, b has bigger f(a), f(b)
        if abs(fa) > abs(fb)
            a,fa = c,fc
        else
            b,fb = c,fc
        end      
        
        # increment iteration counter
        n += 1
    end
    
    # return c as best estimate of root
    c
end

secantsearch (generic function with 2 methods)

## Verify algorithm on a simple test case

Let's try out `secantsearch` on a simple function whose root we know, say, $f(x) = x^2 - 4$, which has a root $x=2$.

In [3]:
f(x) = x^2 - 4

f (generic function with 1 method)

In [4]:
x = secantsearch(f,1.0,4.0,true)

n = 1, c = 1.6, f(c) = -1.4399999999999995
n = 2, c = 2.1538461538461537, f(c) = 0.6390532544378695
n = 3, c = 1.9836065573770492, f(c) = -0.06530502553077122
n = 4, c = 1.9993904297470282, f(c) = -0.00243790943599409
n = 5, c = 2.000002508903471, f(c) = 1.003562017931614e-5
n = 6, c = 1.9999999996176037, f(c) = -1.5295853472707677e-9
n = 7, c = 1.9999999999999998, f(c) = -8.881784197001252e-16


1.9999999999999998

Oh, sweet! It converges from order-1 error to nearly double-precision accuracy in only seven iterations. Should we be concerned that it returns 1.9999999999999998 instead of 2.0 exactly? No, in general, we can't expect to find exact answers with floating-point computations. 

## Plotting convergence

HW1 asks you to modify your `bisectsearch` and `newtonsearch` codes to plot the error versus iteration number. I'll illustrate here how to do that with `secantsearch`.

Main ideas of the revision:
  - Add a vector `x` to store the value of the best guess $x_n$ as you iterate $n=1,2,\ldots$.
  - Return both the final answer `r` (held in variable `c`) and the vector $x$ of iterates.
  - After the function returns, compute the error $e_n = |x_n - r|$ and plot it logarithmically. 
  - Remove the last value of $x$ before plotting so you don't take the log of zero.
  - Take some care to get a sensible value for the first iterate $x_1$, even if the search fails at $n=1$. 



In [5]:
function secantsearch(f, a::AbstractFloat, b::AbstractFloat, diagnostics=false) 
# return solution x of f(x) == 0, with a,b serving as initial guesses
    
    fa = f(a)           # cache values of f(a), f(b)     
    fb = f(b)
    
    # set c to better of a,b (and fc, too)
    c,fc = abs(fa) < abs(fb) ? (a,fa) : (b,fb)
    
    floattype = typeof((a+b)/2) # this'll eval to bigger of a,b types
    ϵ = eps(floattype)  # machine epsilon
    n = 1               # iteration count
    N = 1024            # max iterations
    
    # start vector for storing root iterates xₙ, assign x₁ = c
    x = zeros(floattype, 1)
    x[1] = c
    
    # stop when f(c) is small, |b-a| is small relative to a and b
    # or when we reach maximium number of iterations
    while abs(fc) > 10ϵ && abs(b-a)/(abs(a)+abs(b)) > 10ϵ && n < N
    
        # compute new values for c, fc by secant approximation
        c = a - fa*(b-a)/(fb-fa)
        fc = f(c)
        
        # print diagnostics if they're turned on
        diagnostics && println("n = $n, c = $c, f(c) = $(f(c))")
        
        # error and exit if f(c) is bigger than both f(a) and f(b)
        if abs(fc) > abs(fa) && abs(fc) > abs(fb)
            println("secantsearch error: f(c) is bigger than f(a), f(b)")
            break # break out of while loop, don't store bad value c
        end
        
        # replace whichever of a, b has bigger f(a), f(b)
        if abs(fa) > abs(fb)
            a,fa = c,fc
        else
            b,fb = c,fc
        end      
        
        # store c in x, then increment iteration counter
        if n==1 
            x[1] = c     # write over initial x[1]
        else       
            push!(x,c)   # otherwise save c as x[n]
        end
        n += 1
    end
    
    # return c and vector x
    c, x
    
end

secantsearch (generic function with 2 methods)

#### Run the search again to get the vector of iterates

In [6]:
r,x = secantsearch(f,1.0,4.0,true)

n = 1, c = 1.6, f(c) = -1.4399999999999995
n = 2, c = 2.1538461538461537, f(c) = 0.6390532544378695
n = 3, c = 1.9836065573770492, f(c) = -0.06530502553077122
n = 4, c = 1.9993904297470282, f(c) = -0.00243790943599409
n = 5, c = 2.000002508903471, f(c) = 1.003562017931614e-5
n = 6, c = 1.9999999996176037, f(c) = -1.5295853472707677e-9
n = 7, c = 1.9999999999999998, f(c) = -8.881784197001252e-16


(1.9999999999999998, [1.6, 2.15385, 1.98361, 1.99939, 2.0, 2.0, 2.0])

#### Plot logarithmically,  after removing the last iterate to avoid computing log(0)

In [7]:
pop!(x)

1.9999999999999998

In [8]:
plot(abs.(x-r), marker=:circ, yscale=:log10, label="|x_n - r|")
plot!(xlabel="n", ylabel="|x_n - r|", ylims=(1e-10, 1))

<?xml version="1.0" encoding="utf-8"?>
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 2 
 
 
 4 
 
 
 6 
 
 
 10^-10 
 
 
 10^-8 
 
 
 10^-6 
 
 
 10^-4 
 
 
 10^-2 
 
 
 10^0 
 
 
 n 
 
 
 |x_n - r| 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 |x_n - r|

## Does the convergence match the theoretical estimate of convergence?

The secant method converges at a rate between bisection's linear convergence and Newton's quadratic convergence. Specifially, for twice differentiable functions and guesses sufficiently close to a non-repeated root, the secant method converges as

\begin{equation*}
\lim_{n\rightarrow \infty} \frac{e_{n+1}}{e_n^\alpha} = M
\end{equation*}

where $\alpha = (1+\sqrt{5})/2 \approx 1.62$ and $M$ is a finite constant related to $f$ and its derivatives at the root. 

How can we check the actual convergence at finite $n$ is close to the theoretical estimate in the large-$n$ limit? The easiest way is to compute $e_{n+1}/e_n^\alpha$ for $n=1,2,\dots$ and see if the values are roughly constant. I'll name the vector of errors `err` to avoid a clash with Julia's built-in Euler constant `e`.

In [9]:
err = abs.(x-r)  # vector of errors

6-element Array{Float64,1}:
 0.4        
 0.153846   
 0.0163934  
 0.00060957 
 2.5089e-6  
 3.82396e-10

In [10]:
α = (1+sqrt(5))/2

1.618033988749895

In [11]:
err[2:end]./(err[1:end-1]).^α

5-element Array{Float64,1}:
 0.67759 
 0.338838
 0.471789
 0.399415
 0.440899

Looks like $e_{n+1}\,/\,e_n^\alpha = M \approx  0.44$ (taking the last value as the closest to the large-$n$ limit). Let's construct a theoretical error curve from this formula and the value $e_1 = 0.4$. 

In [12]:
N = length(err)
err_theory = zeros(N)
err_theory[1] = 0.4
for n=2:N
    err_theory[n] = 0.44*err_theory[n-1]^α
end
err_theory

6-element Array{Float64,1}:
 0.4        
 0.0999015  
 0.0105859  
 0.000280152
 7.85622e-7 
 5.83044e-11

Now add the thoeretical error estimate to the plot.

In [13]:
plot!(err_theory, yscale=:log10, linestyle=:dash, label="e_n, theorectical")

<?xml version="1.0" encoding="utf-8"?>
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 2 
 
 
 4 
 
 
 6 
 
 
 10^-10 
 
 
 10^-8 
 
 
 10^-6 
 
 
 10^-4 
 
 
 10^-2 
 
 
 10^0 
 
 
 n 
 
 
 |x_n - r| 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 |x_n - r| 
 
 
 
 e_n, theorectical

Not bad.